core/services/sleeper_task.go

```diff
@@ -47,84 +28,60 @@ type sleeperTask struct {
 // WakeUp does not block.
 //
 func NewSleeperTask(worker Worker) SleeperTask {
-	return &sleeperTask{
-		worker: worker,
-		waker:  make(chan struct{}, 1),
-		closer: make(chan struct{}, 1),
-		closed: make(chan struct{}, 1),
-		state:  unstarted,
-		mutex:  sync.Mutex{},
+	s := &sleeperTask{
+		worker:      worker,
+		chQueue:     make(chan struct{}, 1),
+		chQueueDone: make(chan struct{}),
+		chStopped:   make(chan struct{}),
 	}
-}
 
-// Start begins the SleeperTask
-// Calling start on a started task will return an error
-func (s *sleeperTask) Start() error {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	switch s.state {
-	case stopped:
-		return errors.New("cannot start a sleeper task that has been stopped")
-	case started:
-		return errors.New("sleeper task is already started")
-	}
-	s.state = started
 	go s.workerLoop()
-	return nil
+
+	return s
 }
 
-// Stop stops the SleeperTask
-// It can only be called once
-// It will block until all tasks have completed, or the timeout is reached
-// Stopped tasks cannot be restarted
+// Stop stops the SleeperTask.  It never returns an error.  Its error return
+// exists so as to satisfy other interfaces.
 func (s *sleeperTask) Stop() error {
-	s.mutex.Lock()
-	if s.state == stopped {
-		s.mutex.Unlock()
-		return errors.New("sleeper task is already stopped")
-	}
-	s.state = stopped
-	s.mutex.Unlock()
-
-	s.closer <- struct{}{}
-	// NOTE: Closing the channels will cause the rogue task to panic if it does eventually complete
-	defer close(s.waker)
-	defer close(s.closer)
-	defer close(s.closed)
 	select {
-	case <-s.closed:
-		return nil
-	case <-time.After(stopWaitTime):
-		return fmt.Errorf("task did not complete within %s", stopWaitTime.String())
+	case <-s.chStopped:
+		panic("already stopped")
+	default:
 	}
+
+	close(s.chStopped)
+
+	<-s.chQueueDone
+	return nil
 }
 
 // WakeUp wakes up the sleeper task, asking it to execute its Worker.
-// Idempotent, can be called multiple times but will only wake the worker once (until the worker finishes again)
-func (s *sleeperTask) WakeUp() error {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	if s.state == stopped {
-		return errors.New("cannot wake up stopped sleeper task")
+func (s *sleeperTask) WakeUp() {
+	select {
+	case <-s.chStopped:
+		panic("sleeper task is already stopped")
+	default:
 	}
 
 	select {
-	case s.waker <- struct{}{}:
+	case s.chQueue <- struct{}{}:
 	default:
 	}
-	return nil
 }
 
-// workerLoop is the goroutine behind the sleeper task that waits for a signal
-// before kicking off the worker
 func (s *sleeperTask) workerLoop() {
+	defer close(s.chQueueDone)
+
 	for {
 		select {
-		case <-s.waker:
+		case <-s.chQueue:
 			s.worker.Work()
-		case <-s.closer:
-			s.closed <- struct{}{}
-			return
+		default:
+			select {
+			case <-s.chStopped:
```

Comment:
Can this just go up one level?

Reply:
My intention here was to make sure that the worker cleared out the queue before stopping.  This probably doesn't matter in production, but it caused a test or two to fail because `Stop` was called directly after `WakeUp`, causing the worker to never run.  The alternative is to add a `time.Sleep` in those tests before `Stop`.  What's your preference?

Reply:
It's absolutely essential to clear out the worker queue before stopping. That's the entire point of this original work, because CI fails if we don't do that.

I added a safety hatch in my original version with the timeout, so if the task doesnt stop within a certain time we return an error and/or panic otherwise it could potentially block forever if the task doesn't finish.